kernel/src/cpu/ipi.rs

```diff
@@ -364,16 +364,22 @@ pub fn send_ipi(
             }
         }
         _ => {
+            let mut target_count: usize = 0;
             for cpu in PERCPU_AREAS.iter() {
-                ipi_board.pending.fetch_add(1, Ordering::Relaxed);
-                cpu.as_cpu_ref().ipi_from(sender_cpu_index);
+                // Ignore the current CPU and CPUs that are not online.
+                let cpu_shared = cpu.as_cpu_ref();
+                if cpu_shared.is_online() && cpu_shared.apic_id() != this_cpu().get_apic_id() {
+                    target_count += 1;
+                    cpu_shared.ipi_from(sender_cpu_index);
+                }
```

Comment:
Would something like this be a bit cleaner?
```suggestion
            for cpu in PERCPU_AREAS
                .iter()
                .map(|c| c.as_cpu_ref())
                .filter(|c| c.is_online() && c.apic_id() != this_cpu().get_apic_id())
            {
                target_count += 1;
                cpu.ipi_from(sender_cpu_index);
            }
```

Reply:
That's a great question.  I don't know how `map` and `filter` affect stack consumption and iterator performance.  It's easy for me to imagine that both `map` and `filter` cause the compiler to generate separate iterator objects, which implies that the loop iteration may occur multiple times (once for each iterator) and that stack consumption grows to accommodate the new iterator objects.  Why take the risk of such inefficiencies for such a small change in readability?

Reply:
I don't have issues with this change the way it is. When I was working on a different project I was asked numerous times by the Rust reviewers to avoid C-like coding styles like this (which is natural to me as a C programmer) and was constantly encouraged to use iterators instead. I'm not sure if the Rust compiler does anything to address your stack consumption concerns. Since I'm not the enforcer of how Rusty this project should be I'll just leave it at that.

Reply:
In my mind, the two things that matter more than anything in a code base are how maintainable the code is, and how well it performs.  I don't personally see a connection between deep embrace of a language's peculiar syntax and maintainability - as long as someone reading the code can clearly understand what it does and can modify it if necessary, then we've met the maintainability goal.  I would be all for embracing Rust-specific iteration constructs if it made the code more maintainable but I don't think it matter much in this case, just as you are saying.  Rust for the sake of Rust is not very compelling by itself; Rust for the sake of safety and maintainability makes much more sense.
 
If I had more time, I would look at the optimized code generation for both examples to see which one is more efficient and would go with that, but I think that will have to be left as an exercise for a future date.  My AI buddy tells me that code comparisons within the `foreach` loop are likely to be more efficient, but who knows if that's actually the truth.